### PR TITLE
I293 no-u-turn sampler

### LINKED ISSUE #293: Implement No-U-Turn Sampler (NUTS)
This is actually the algorithm (not HMC, although they are similar) that powers Stan and Pymc3. Look at the latest papers on Stan/Pymc3 to see what variants they implement.

### PR DESCRIPTION
**Pull request by martinjrobins**
Fixes #293 

- Implemented NUTS sampler as a coroutine
- added ability to send a `None` to logging
- added NUTS notebook
- conditionally import NUTS sampler, only implemented for python version > 3.3

@MichaelClerx, when you have time can you have a look at this and see if it's ok. It's mostly complete, I just have to do the conditional importing. In particular, perhaps I should be using the "initial_phase" functionality of the samplers for the dual averaging?
 
@ben: NUTS works great on logistic problem (my problems yesterday were mostly dumb coding mistakes!)

### CODE REVIEW

File: pints/_logger.py [hunk 1]
```diff
@@ -386,19 +386,22 @@ def log(self, *data):
...
-                    x += ' ' * (width - len(x))
-                elif dtype == _TIME:
-                    x = self._format_time(next(column))
-                elif dtype == _TEXT:
-                    x = str(next(column))[:width]
-                    x += ' ' * (width - len(x))
+                v = next(column)
+                if v is None:
+                    x = ' ' * width
                 else:
```
Comment (MichaelClerx) [inline pints/_logger.py]: Could be an elif in case we want to save space :D

File: pints/_mcmc/_nuts.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import asyncio
+import pints
+import numpy as np
+
+
+class NutsState:
+    """
+    Class to hold information about the current state of the NUTS hamiltonian
```
Comment (MichaelClerx) [inline pints/_mcmc/_nuts.py]: Could have `.` at the end of every sentence?

File: pints/_mcmc/_nuts.py [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        self.s = s
+        self.theta = theta
+        self.L = L
+        self.grad_L = grad_L
+        self.alpha = alpha
+        self.n_alpha = n_alpha
+
+    def update(self, other_state, direction, root):
+        """
+        if root == True, this combines a depth j subtree (`self`) with a depth j+1
```
Comment (MichaelClerx) [inline pints/_mcmc/_nuts.py]: ` ``root == True`` ` etc (double ticks in RST)

File: pints/_mcmc/_nuts.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        # accumulated number of accepted points
+        self.n += n_dash
+
+        # test if the path has done a U-Turn
+        self.s *= s_dash
+        self.s *= int((self.theta_plus - self.theta_minus).dot(self.r_minus) >= 0)
+        self.s *= int((self.theta_plus - self.theta_minus).dot(self.r_plus) >= 0)
+
+
+@asyncio.coroutine
```
Comment (MichaelClerx) [inline pints/_mcmc/_nuts.py]: Could we have some comments in here explaining why these need to be coroutines?

File: pints/_mcmc/_nuts.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+def leapfrog(theta, L, grad_L, r, epsilon, step_size):
+    """ performs a leapfrog step with step size `epsilon*step_size """
```
Comment (MichaelClerx) [inline pints/_mcmc/_nuts.py]: `Performs` and ` ``epsilon*step_size``. `

File: pints/tests/test_mcmc_nuts.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        self.assertEqual(mcmc.delta(), 0.1)
+        self.assertEqual(mcmc.number_adaption_steps(), 2)
+        self.assertEqual(mcmc.leapfrog_step_size()[0], 2)
+        mcmc.set_hyper_parameters([0.1, 2, [4.0, 3.0]])
+        self.assertEqual(mcmc.leapfrog_step_size()[0], 4.0)
+        self.assertEqual(mcmc.leapfrog_step_size()[1], 3.0)
+
+
+if __name__ == '__main__':
+    print('Add -v for more debug output')
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pints/tests/test_mcmc_nuts.py]: done

File: pints/tests/test_mcmc_nuts.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+#
+import unittest
+import numpy as np
+
+import pints
+import pints.toy
+
+from shared import StreamCapture
+
+debug = False
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pints/tests/test_mcmc_nuts.py]: done

File: pints/_mcmc/_nuts.py (hunk 5 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: done

File: pints/_mcmc/_nuts.py (hunk 4 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: done

File: pints/_mcmc/_nuts.py (hunk 3 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: done

File: pints/_mcmc/_nuts.py (hunk 2 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: done

File: pints/tests/test_mcmc_nuts.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+#
+# Tests the basic methods of the Hamiltonian MCMC routine.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pints/tests/test_mcmc_nuts.py]: done

File: pints/_logger.py (hunk 1 above)
Comment (martinjrobins) [reply pints/_logger.py]: done

File: examples/README.md [hunk 9]
```diff
@@ -51,6 +51,7 @@ relevant code.
 
 ### MCMC with gradients
 - [Hamiltonian MCMC](./sampling-hamiltonian-mcmc.ipynb)
+- [No-U-Turn MCMC](./sampling-nuts-mcmc.ipynb)
```
Comment (ben18785) [inline examples/README.md]: Very minor but could we keep this alphabetical.

**Comment by martinjrobins** [comment]
thanks @ben18785, we don't track divergent iterations for nuts, but can add this in a similar way to hamiltonianmcmc. I'll alter the notebook as you suggest as well

**Comment by ben18785** [comment]
Cool -- yep, would be good to track these I think. Ta!

File: pints/_mcmc/_nuts.py [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+        # use slice sampling
+        log_u = np.log(np.random.uniform(0, 1)) + hamiltonian0
+
+        # create initial integration path state
+        state = NutsState(theta, r0, L, grad_L, 1, 1, None, None)
+        j = 0
+
+        # build up an integration path with 2^j points, stopping when we either
+        # encounter a U-Turn, or reach a max number of points 2^10
+        while j < 10 and state.s == 1:
```
Comment (ben18785) [inline pints/_mcmc/_nuts.py]: @martinjrobins Whilst I remember, can we make the tree depth a hyper-parameter?

File: examples/README.md (hunk 9 above)
Comment (martinjrobins) [reply examples/README.md]: done

File: pints/_mcmc/_nuts.py (hunk 10 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: done

**Comment by MichaelClerx** [comment]
Tantalisingly close to being mergeable :D !

**Comment by martinjrobins** [comment]
Almost! I'm just adding the multinomial sampling that they use in Stan

**Comment by MichaelClerx** [comment]
Cool! I just added it to the docs and fixed the test skipping. Hope that doesn't interfere

File: pints/_mcmc/_nuts.py [hunk 11]
```diff
@@ -0,0 +1,719 @@
...
+
+        # next step
+        m += 1
+
+
+class NoUTurnMCMC(pints.SingleChainMCMC):
+    r"""
+
+    Implements No U-Turn Sampler (NUTS) with dual averaging, as described in
+    Algorithm 6 in [1]_.
```
Comment (ben18785) [inline pints/_mcmc/_nuts.py]: Cool. Can we add a note to say that it either does Algo 6 in [1] or does multinomial sampling. In the latter case, can we include pseudocode or something (seeing as it's not written anywhere in Stan!).

**Comment by martinjrobins** [comment]
So I've implemented the multinomial sampling, dense or diagonal mass matrix, and full adaption of epsilon and mass matrix. The multinomial sampling seems to be slightly worse than the slice sampling, so I'm thinking perhaps I've implemented it wrong.....

Note for future: I have split off the adaption of epsilon and mass matrix to a separate class, so this can be used in HMC as well later on

File: pints/_mcmc/_nuts.py (hunk 11 above)
Comment (martinjrobins) [reply pints/_mcmc/_nuts.py]: where should the pseudocode go? In the docstring, can this be formatted nicely?

**Comment by martinjrobins** [comment]
closing this as development of nuts is continuing on `nuts-comparison` branch

**Comment by MichaelClerx** [comment]
Hey @martinjrobins ! Is it OK to delete this branch now?

**Comment by MichaelClerx** [comment]
0 new commits compared to master, I see, so closing!

**Pull request closed without merging**